### PR TITLE
HDDS-8977. Ratis crash if a lot of directories deleted at once

### PR DESCRIPTION
**Pull request by sumitagrawl**
## What changes were proposed in this pull request?

- change max limit for DirectoryDeletingService to 6000 (from 10000), considering, 4KB size (considering acls, key/file name, and other meata)  * 6000, resulting 24MB
- If overall request size crosses the size limit, and excluding the latest request do not have any information to delete, then consider limit as 1000 for the task for deletion and retry with same.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-8977

## How was this patch tested?

- Testcase is added to verify the case of crossing limit size and retrying with 1000 limit for the task.

### CODE REVIEW

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/SnapshotDeletingService.java [hunk 1]
```diff
@@ -89,6 +91,7 @@ public class SnapshotDeletingService extends AbstractKeyDeletingService {
   // from the same table and can send deletion requests for same snapshot
   // multiple times.
   private static final int SNAPSHOT_DELETING_CORE_POOL_SIZE = 1;
+  private static final int MIN_ERR_LIMIT_PER_TASK = 1000;
```
Comment (sadanand48) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/SnapshotDeletingService.java:94]: Are we sure that 1000 entries would be well under the Ratis Byte limit or should we calculate this based on the remainingBufLimit?

File: hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/service/TestDirectoryDeletingService.java [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+        .setBucketName(bucketName)
+        .setKeyName("dir" + longName)
+        .setReplicationConfig(StandaloneReplicationConfig.getInstance(
+            HddsProtos.ReplicationFactor.ONE))
+        .setDataSize(0).setRecursive(true)
+        .build();
+    writeClient.deleteKey(delArgs);
+
+    // wait for file count of only 100 but not all 2000 files
+    // as logic, will take 100 files
```
Comment (sadanand48) [inline hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/service/TestDirectoryDeletingService.java:150]: 1000 files?

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/SnapshotDeletingService.java (hunk 1 above)
Comment (sumitagrawl) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/SnapshotDeletingService.java:94]: 32MB is limit set default, but even 6000* (4KB as approx size taken) => 24MB crosses, then we are going with 1000, i.e. 32MB/1000 => 32KB.
We can not do calculation for size till its serialize, and will have impact like binary calculation till we achieve the meeting criteria. This kind of thing is not required, and even if 1000 entry, limit is not met, user need increase buffer size of Ratis.

So this logic is not done, as its not simple to identify remaining buf directly without serializing.

File: hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/service/TestDirectoryDeletingService.java (hunk 2 above)
Comment (sumitagrawl) [reply hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/service/TestDirectoryDeletingService.java]: done

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/SnapshotDeletingService.java (hunk 1 above)
Comment (sadanand48) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/SnapshotDeletingService.java:94]: Thanks for explaining.

**Comment by errose28** [comment]
Thanks for working on this @sumitagrawl. I did not get a chance to thoroughly review the changes, but checking the size of the proto is a good idea. Looks like there should not be any performance issues with `getSerializedSize()` according to this old [mail thread](https://groups.google.com/g/protobuf/c/BwzA9jIigZU?pli=1) I dug up. Would it make sense to increase the Ratis buffer limit in the PR as well? After RATIS-1783 that change should be effective.

cc @kerneltime @szetszwo who also looked at this issue in the past.

**Comment by sumitagrawl** [comment]
> Would it make sense to increase the Ratis buffer limit in the PR as well? After RATIS-1783 that change should be effective.

@errose28 
IMO, increase ratis buffer size needs to be handled separately, as it involves,
1. performance impact verification
2. compatibility if rolling upgrade supported
3. involves changes for OM, SCM, DN as common way of handling.

**Comment by sadanand48** [comment]
Thanks @sumitagrawl for the patch, @errose28 for the review/comments.

**Pull request merged**